packages/flutter/lib/src/widgets/text_formatter.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/services.dart';
+
+/// A [TextInputFormatter] can be optionally injected into an [EditableText] 
+/// to provide as-you-type validation and formatting of the text being edited.
+/// 
+/// Text modification should only be applied when text is being committed by the
+/// IME and not on text under composition (i.e. when 
+/// [TextEditingValue.composing] is collapsed). 
+/// 
+/// A concrete [BlacklistingTextInputFormatter] which removes blacklisted 
+/// characters upon edit commit is provided. 
+/// 
+/// To create custom formatters, extend the [TextInputFormatter] class and 
+/// implement the [formatEditUpdate] method.
+/// 
+/// See also:
+///  
+///  * [EditableText] on which the formatting apply.
+///  * [BlacklistingTextInputFormatter], a provided formatter for blacklisting
+///    characters.
+abstract class TextInputFormatter {
+  /// Called when text is being typed or cut/copy/pasted in the [EditableText].
+  /// 
+  /// You can override the resulting text based on the previous text value and
+  /// the incoming new text value.
+  TextEditingValue formatEditUpdate(
+    TextEditingValue oldValue, 
+    TextEditingValue newValue
+  );
+}
+
+/// A [TextInputFormatter] that prevents the insertion of blacklisted.
+/// 
+/// Instances of blacklisted characters found in the new [TextEditingValue]s 
+/// will be replaced with the [replacementString] which defaults to ``.
+/// 
+/// It attempts to preserve the existing [TextEditingValue.selection] to 
+/// sensible values. 
```

Comment:
Maybe: 
```
/// This formatter attempts to preserve the [TextEditingValue.selection].
```

That said, as-is, I'm not convinced this adds much. It feels like we're saying 'This formatter behaves how you'd expect it to.' Perhaps more directly state that this formatter will adjust the selection in the case that blacklisted characters were removed?